Unify usage of indent_style

Triton171 opened the pull request:
Previously, we used sometimes `doc.indent_style` and sometimes `doc.language_config.indent`. This caused some issues if they differ (e.g. if a non-default indent style is auto-detected in a document or manually set). For example, in a markdown file, setting `:indent-style 8` and then continuously pressing enter (starting on an indented line) increases the indent each time. This happens because the indent_style used to detect the indentation level of an existing line is different from the indent_style used to convert an indentation level to a string. The same issue occurs in any language without indent queries.
 
With these changes, `language_config.indent` is only used to initialize `doc.indent_style` and afterwards ignored. Resolves #2236 and #3406.

pascalkuthe reviewed (changes requested):
If I understand this change correctly that means that the number of spaces in the indent unit now always corresponds to the tabwidth.

In other words:
* If my document contains both spaces and tabs
* I set the indent width to 4
* Then all tabs are rendered with a width of 4

This does not correspond to the current config whichs allows setting `tab-width`  separately from the `unit` field.

So for example this change would break using 4 spaces for indentation and a `tab-width` of two:

```
indent-style = { unit = "    ", tab-width=2 }
```

Is unifying these two really necessary to fix the underling issue here?

From my understanding, the correct fix would be to have to separate parameters for `indent_level_for_line`
the `tab-width` which is used when `\t` is encoutered and the `indent-width` wich is used for calulating the indent level

Triton171 commented:
Thanks for the super-fast review! You're right, I hadn't really thought about the distinction between `tab_width` (the width the tab character is rendered at) and `indent_width` (the width of an indentation level). It appears that I'm not the only one, as a significant percentage of the usages of `tab_width` in the current master should actually be `indent_width`. I've now made the difference a bit clearer in the code and fixed these bugs.

I also removed the `doc.tab_width()` function and replaced their usages by `doc.indent_style.tab_width`. This hopefully makes it more obvious that there are 2 different widths to consider when computing indentation.

pascalkuthe commented on helix-view/src/document.rs:
```diff
@@ -1122,13 +1130,6 @@ impl Document {
         self.syntax.as_ref()
     }
 
```
With this config removed `tab-width` is not updated on `:config-reload` anymore. I actually think that we should keep this function around and not make `IndentStyle` a struct at all. 

Rather we should add a second function to the document to compute the `indent_width`:
``` suggestion
     /// Tab size in columns.
    pub fn tab_width(&self) -> usize {
        self.language_config()
            .and_then(|config| config.indent.as_ref())
            .map_or(4, |config| config.tab_width) // fallback to 4 columns
     }

    /// Indentation size in columns
    pub fn indent_width(&self) -> usize {
        match self.indent_style{
            IndentStyle::Tab => self.tab_width(),
            IndentStyle::Spaces(width) => width as usize
         }
    }
```

pascalkuthe reviewed (changes requested):
Yeah it's definitely a bit subtle :+1: People usually stick to a single indent style they tend to o think about `indent_width` and `tab_width` as the same thing. 

The changes to the autodetection now LGTM :+1: 

However this PR regresses some things like config reloading. I think a simpler fix is possible by simply adding a new `indent_width` function instead of actually changing `IndentStyle`. I added suggestion below.

pascalkuthe reviewed (approved):
Thanks for tackling this, LGTM now :+1: 

Ident width and tab width is a subtle but important distinction, great to see that cleared up

Triton171 commented:
Thanks for all the feedback!
